### DISCUSSION TITLE
Introduce an annotation for overridability of GDScript functions

### DISCUSSION THREAD

**Discussion by Rob23oba** [godotengine/godot-proposals#10155] [Scripting]
Currently all GDScript functions can be overridden at will. I however propose that in the future, GDScript functions that should be overridable need to be marked with `@virtual` or a similar annotation, e.g. previously:
```gdscript
class MyClass:
	func some_animation(from: float, to: float, t: float) -> float:
		return lerpf(from, to, t)
class SubClass:
	extends MyClass

	func some_animation(from: float, to: float, t: float) -> float:
		return lerpf(from, to, t * t)
```
Now:
```gdscript
class MyClass:
	@virtual
	func some_animation(from: float, to: float, t: float) -> float:
		return lerpf(from, to, t)
class SubClass:
	extends MyClass

	func some_animation(from: float, to: float, t: float) -> float:
		return lerpf(from, to, t * t)
```
In this example, the class `MyClass` defined a simple linear interpolation function. However, it was overridden by `SubClass` to turn it into a quadratic interpolation. When looking at just the code of `MyClass` however, it is not obvious that this function is being overridden at all - one could think that the definition in `MyClass` is the one that is actually being used, however due to it being overridden this might not be the case. Introducing an `@virtual` annotation would solve this problem since it is now clearly documented that, indeed, that function can be overridden and one can now search such occurrences.
This change could also benefit performance, though: the GDScript compiler doesn't optimize very much currently but with this change it would be possible to inline short functions or at least replace function calls with direct references to the code.
Additionally, this makes GDScript functions similar to builtin-in functions that are also only overridable when they are defined as virtual.
However there is one major downside to this: all existing functions that can are being overridden would need to now be defined as `@virtual`. To simplify this, one could introduce some kind of "quick fix" tool that automatically adds `@virtual` where needed.

**Comment by AThousandShips** [comment]
Thank you for your proposal, consolidating in:
* https://github.com/godotengine/godot-proposals/issues/1631

And a few other related proposals already open
